src/gpl/src/nesterovBase.cpp

```diff
@@ -1934,79 +1819,99 @@ float NesterovBase::getStepLength(
-  // wireLengthGradSum_, densityGradSum_, gradSum)
-  for (size_t i = 0; i < gCells_.size(); i++) {
-    GCell* gCell = gCells_.at(i);
-    wireLengthGrads[i]
-        = nbc_->getWireLengthGradientWA(gCell, wlCoeffX, wlCoeffY);
-    densityGrads[i] = getDensityGradient(gCell);
-
-    // Different compiler has different results on the following formula.
-    // e.g. wireLengthGradSum_ += fabs(~~.x) + fabs(~~.y);
-    //
-    // To prevent instability problem,
-    // I partitioned the fabs(~~.x) + fabs(~~.y) as two terms.
-    //
-    wireLengthGradSum_ += std::fabs(wireLengthGrads[i].x);
-    wireLengthGradSum_ += std::fabs(wireLengthGrads[i].y);
+  for (auto& gCell : newGCells_) {
+    GCellState& state = gCell->state;
+    state.*wireLengthGradsPtr
+        = nbc_->getWireLengthGradientWA(gCell.get(), wlCoeffX, wlCoeffY);
+    state.*densityGradsPtr = getDensityGradient(gCell.get());
 
-    densityGradSum_ += std::fabs(densityGrads[i].x);
-    densityGradSum_ += std::fabs(densityGrads[i].y);
+    wireLengthGradSum_ += std::fabs((state.*wireLengthGradsPtr).x);
+    wireLengthGradSum_ += std::fabs((state.*wireLengthGradsPtr).y);
 
-    sumGrads[i].x = wireLengthGrads[i].x + densityPenalty_ * densityGrads[i].x;
-    sumGrads[i].y = wireLengthGrads[i].y + densityPenalty_ * densityGrads[i].y;
+    densityGradSum_ += std::fabs((state.*densityGradsPtr).x);
+    densityGradSum_ += std::fabs((state.*densityGradsPtr).y);
 
-    FloatPoint wireLengthPreCondi = nbc_->getWireLengthPreconditioner(gCell);
-    FloatPoint densityPrecondi = getDensityPreconditioner(gCell);
+    state.*sumGradsPtr
+        = FloatPoint((state.*wireLengthGradsPtr).x
+                         + densityPenalty_ * (state.*densityGradsPtr).x,
+                     (state.*wireLengthGradsPtr).y
+                         + densityPenalty_ * (state.*densityGradsPtr).y);
+
+    FloatPoint wireLengthPreCondi
+        = nbc_->getWireLengthPreconditioner(gCell.get());
+    FloatPoint densityPrecondi = getDensityPreconditioner(gCell.get());
 
     FloatPoint sumPrecondi(
         wireLengthPreCondi.x + densityPenalty_ * densityPrecondi.x,
         wireLengthPreCondi.y + densityPenalty_ * densityPrecondi.y);
 
-    if (sumPrecondi.x <= npVars_->minPreconditioner) {
-      sumPrecondi.x = npVars_->minPreconditioner;
-    }
-
-    if (sumPrecondi.y <= npVars_->minPreconditioner) {
-      sumPrecondi.y = npVars_->minPreconditioner;
+    (state.*sumGradsPtr).x
+        = (sumPrecondi.x <= npVars_->minPreconditioner)
+              ? (state.*sumGradsPtr).x / npVars_->minPreconditioner
+              : (state.*sumGradsPtr).x / sumPrecondi.x;
+    (state.*sumGradsPtr).y
+        = (sumPrecondi.y <= npVars_->minPreconditioner)
+              ? (state.*sumGradsPtr).y / npVars_->minPreconditioner
+              : (state.*sumGradsPtr).y / sumPrecondi.y;
+
+    gradSum += std::fabs((state.*sumGradsPtr).x)
+               + std::fabs((state.*sumGradsPtr).y);
```

Comment:
I think this code would be a lot clearer if you only dereferenced state.*sumGradsPtr once and then stored it in a local variable